skiko/src/wasmJsMain/kotlin/org/jetbrains/skia/impl/Native.wasm.kt

```diff
@@ -0,0 +1,408 @@
+    var address = dest
+    for (value in src) {
+        skia_memSetChar(address, value)
+        address += Char.SIZE_BYTES
+    }
+}
+
+internal actual fun toWasm(dest: NativePointer, src: IntArray) {
+    var address = dest
+    for (value in src) {
+        skia_memSetInt(address, value)
+        address += Int.SIZE_BYTES
+    }
+}
+
+internal actual fun toWasm(dest: NativePointer, src: FloatArray) {
+    var address = dest
+    for (value in src) {
+        skia_memSetFloat(address, value)
+        address += Float.SIZE_BYTES
+    }
+}
+
+internal actual fun toWasm(dest: NativePointer, src: DoubleArray) {
+    var address = dest
+    for (value in src) {
+        skia_memSetDouble(address, value)
+        address += Double.SIZE_BYTES
+    }
+}
+
+internal actual fun fromWasm(src: NativePointer, result: ByteArray) {
+    var address = src
+    for (index in result.indices) {
+        result[index] = skia_memGetByte(address)
+        address += Byte.SIZE_BYTES
+    }
+}
+
+internal actual fun fromWasm(src: NativePointer, result: ShortArray) {
+    var address = src
+    for (index in result.indices) {
+        result[index] = skia_memGetShort(address)
+        address += Short.SIZE_BYTES
+    }
+}
+
+internal actual fun fromWasm(src: NativePointer, result: IntArray) {
+    var address = src
+    for (index in result.indices) {
+        result[index] = skia_memGetInt(address)
+        address += Int.SIZE_BYTES
+    }
+}
+
+internal actual fun fromWasm(src: NativePointer, result: FloatArray) {
+    var address = src
+    for (index in result.indices) {
+        result[index] = skia_memGetFloat(address)
+        address += Float.SIZE_BYTES
+    }
+}
+
+internal actual fun fromWasm(src: NativePointer, result: DoubleArray) {
+    var address = src
+    for (index in result.indices) {
+        result[index] = skia_memGetDouble(address)
+        address += Double.SIZE_BYTES
+    }
+}
+
+internal actual fun stringToUTF8(str: String, outPtr: NativePointer, maxBytesToWrite: Int) {
+    if (maxBytesToWrite <= 0) return
+
+    val utf8 = str.encodeToByteArray()
+    val lastIndex = minOf(maxBytesToWrite - 1, utf8.size)
+
+    var index = 0
+    while (index < lastIndex) {
+        skia_memSetByte(outPtr + index, utf8[index])
+        index++
+    }
+    skia_memSetByte(outPtr + index, 0)
+}
+
+internal actual class InteropScope actual constructor() {
+    private val elements = mutableListOf<NativePointer>()
+    private var callbacksInitialized = false
+
+    actual fun toInterop(string: String?): InteropPointer {
+        return if (string != null) {
+            val data = _malloc(string.length * 4)
+            stringToUTF8(string, data, string.length * 4)
+            elements.add(data)
+            data
+        } else {
+            0
+        }
+    }
+
+    private fun toInterop(array: ByteArray?, copyArrayToWasm: Boolean): InteropPointer {
+        return if (array != null && array.isNotEmpty()) {
```

Comment:
Would we need to create a native array, even if it is empty? Null and empty might be handled differently.